akka-actor-tests/src/test/scala/akka/actor/CoordinatedShutdownSpec.scala

```diff
@@ -150,8 +149,96 @@ class CoordinatedShutdownSpec
         testActor ! "C"
         Future.successful(Done)
       }
-      Await.result(co.run(UnknownReason), remainingOrDefault)
-      receiveN(4) should ===(List("A", "B", "B", "C"))
+      whenReady(co.run(UnknownReason)) { _ =>
+        receiveN(4) should ===(List("A", "B", "B", "C"))
+      }
+    }
+
+    "cancel tasks" in {
+      import system.dispatcher
+      val phases = Map("a" -> emptyPhase, "b" -> phase("a"))
+      val co = new CoordinatedShutdown(extSys, phases)
+
+      val task0Counter = new AtomicInteger()
+
+      val task0: () => Future[Done] = () =>
+        Future {
+          task0Counter.incrementAndGet()
+          Done
+        }
+
+      co.addCancellableTask("a", "task0-copy0")(task0)
+      co.addCancellableTask("a", "task0-copy1")(task0)
+      val cancellable0 = co.addCancellableTask("a", "task0-copy2")(task0)
+
+      object Task1 { // We can also add a method as a task
+        val counter = new AtomicInteger()
+
+        private def taskMethod(): Future[Done] = Future {
+          counter.incrementAndGet()
+          Done
+        }
+
+        def register(phase: String, name: String): Cancellable = co.addCancellableTask(phase, name)(taskMethod _)
+      }
+
+      Task1.register("b", "task1-copy0")
+      Task1.register("b", "task1-copy1")
+      val cancellable1 = Task1.register("b", "task1-copy2")
+
+      // Adding the same task twice under the same phase/name will still run it twice
```

Comment:
I'd much prefer if these comments were instead separate test cases (with the comments as names then) rather than one long test covering different aspects.

Reply:
Good idea.